drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py

```diff
@@ -0,0 +1,315 @@
+
+        dtoa = [get_time_delta(t_upsamp, non_ref, signals[0]) \
+                for non_ref in signals_upsamp[1 : self.receiver_count]]
+
+        return dtoa
+
+    #Passive Sonar Services
+
+    @thread_lock(lock)
+    def get_pulse_heading(self, srv):
+        '''
+        Returns the heading towards an active pinger emmiting at <self.target_freq>.
+        Heading will be a unit vector in hydrophone_array frame
+        '''
+        self.ser.flushInput()
+        try:
+            # Request raw signal tx until start bit is received
+            readin = None
+            timeout_ref = rospy.Time.now()
+            while readin == None or ord(readin) != ord(self.TX_START):
+                self.ser.write(self.TX_REQUEST)
+                readin = self.ser.read(1)
+                if readin == '':  # serial read timed out
+                    return {}
+        except serial.SerialTimeoutException as e:
+            print e
+            return {}
+        except serial.SerialException as e:
+            print  e
+            return {}
+
+        time = rospy.Time.now()
+        signals = receive_signals(self.ser, self.receiver_count, self.signal_sz, 
+                                  self.scalar_sz, self.signal_bias)
+        heading = self.multilaterator.getPulseLocation(self.get_dtoa(signals))
+
+        # Add heaing observation to buffers if the signals are above the variance threshold
+        variance = np.var(signals)
+        if variance > self.min_variance:
+            p0, R = getHydrophonePose(time)
+            R = tf.transformations.quaternion_matrix(R)
+            map_offset = R.dot(heading)
+            p1 = p0 + map_offset
+
+            self.visualize_heading(p0, p1, bgra=[1.0, 0, 0, 0.50], length=2.0)
+
+            self.heading_start = np.append(self.heading_start, p0, axis=0)
+            self.heading_end = np.append(self.heading_end, p1, axis=0)
+            self.observation_variances = np.append(self.observation_variances, variance)
+
+            # delete softest samples if we have over max_observations
+            if len(self.line_array) >= self.observation_buffer_size:
+                softest_idx = np.argmin(self.observation_variances)
+                self.heading_start = np.delete(self.line_array, softest_idx, axis=0)
+                self.heading_end = np.delete(self.line_array, softest_idx, axis=0)
+                self.observation_variances = np.delete(self.observation_variances, softest_idx, axis=0)
+
+        header = make_header(frame=self.locating_frame)
+        heading = {'x' : heading[0], 'y' : heading[1],  'z' : heading[2]}
+        return {'header' : header, 'heading' : heading}
+
+    def estimate_pinger_position(self, req):
+        '''
+        Uses a buffer of prior observations to estimate the position of the pinger as the intersection
+        of a set of 3d lines in the least-squares sense.
+        '''
+        assert len(self.heading_start) > 1
```

Comment:
Will this assertion cause the script to crash? If so it'd probably be favorable to just return a blank response with 0 observations